// @abacatepay/sdk/dist/index.js
// ../rest/src/errors.ts
class AbacatePayError extends Error {
  message;
  constructor(message) {
    super(message);
    this.message = message;
    this.name = "AbacatePayError";
  }
}

class HTTPError extends Error {
  message;
  route;
  status;
  method;
  constructor(message, route, status, method) {
    super(message);
    this.message = message;
    this.route = route;
    this.status = status;
    this.method = method;
    this.name = `HTTPError(${route})`;
  }
}

class TimeoutError extends Error {
  message;
  timeout;
  constructor(message, timeout) {
    super(message);
    this.message = message;
    this.timeout = timeout;
    this.name = "TimeoutError";
  }
}

// ../rest/src/utils.ts
var sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
var RATE_LIMIT_STATUS_CODE = 429;
var RETRYABLE_STATUS = [
  408,
  425,
  RATE_LIMIT_STATUS_CODE,
  500,
  502,
  503,
  504
];
var BASE_DELAY_MS = 300;
var MAX_DELAY_MS = 1e4;
var backoff = (attempt) => {
  const exp = Math.min(MAX_DELAY_MS, BASE_DELAY_MS * 2 ** attempt);
  const jitter = Math.random() * exp * 0.3;
  return Math.floor(exp + jitter);
};

// ../rest/src/client.ts
var FIVE_SEC_IN_MS = 5000;

class REST {
  options;
  constructor(options = {}) {
    this.options = options;
  }
  setSecret(secret) {
    this.options.secret = secret;
    return this;
  }
  get(route, options) {
    return this.makeRequest(route, { ...options, method: "GET" });
  }
  post(route, options) {
    return this.makeRequest(route, { ...options, method: "POST" });
  }
  delete(route, options) {
    return this.makeRequest(route, { ...options, method: "DELETE" });
  }
  put(route, options) {
    return this.makeRequest(route, { ...options, method: "PUT" });
  }
  patch(route, options) {
    return this.makeRequest(route, { ...options, method: "PATCH" });
  }
  async makeRequest(route, options, attempt = 0) {
    const url = this.makeURL(route, options.query);
    const timeout = this.options.timeout ?? FIVE_SEC_IN_MS;
    try {
      const response = await fetch(url, {
        method: options.method,
        signal: AbortSignal.timeout(timeout),
        headers: this.makeHeaders(options.headers),
        body: "body" in options ? JSON.stringify(options.body) : null
      });
      if (!response.ok)
        return this.handleError({ route, attempt, options, response });
      return this.process(response);
    } catch (err) {
      const isTimeoutError = err?.name === "TimeoutError";
      if (isTimeoutError)
        throw new TimeoutError(`Your request timed out (Waited for ${timeout}ms)`, timeout);
      throw new HTTPError(`${err}`, route, 0, "");
    }
  }
  async handleError({
    route,
    options,
    attempt,
    response
  }) {
    if (RETRYABLE_STATUS.includes(response.status)) {
      const { onRateLimit, retry = options.retry ?? { max: 3 } } = this.options;
      if (attempt >= retry.max)
        throw new HTTPError(`${retry.max} attempts were performed, all failed`, route, response.status, options.method);
      if (retry.onRetry)
        await retry.onRetry({ attempt, options, response });
      if (response.status === RATE_LIMIT_STATUS_CODE && onRateLimit)
        await onRateLimit(response);
      const delay = (retry.backoff ?? backoff)(attempt);
      await sleep(delay);
      return this.makeRequest(route, options, attempt + 1);
    }
    const { error } = await response.json();
    throw new AbacatePayError(error);
  }
  async process(response) {
    const NO_CONTENT_STATUS_CODE = 204;
    if (response.status === NO_CONTENT_STATUS_CODE)
      return;
    const { data, error } = await response.json();
    if (error)
      throw new AbacatePayError(error);
    return data;
  }
  makeURL(route, query) {
    const base = `${this.options.base ?? "https://api.abacatepay.com/v"}${this.options.version ?? 1}${route}`;
    return query ? `${base}?${new URLSearchParams(query)}` : base;
  }
  makeHeaders(custom) {
    const {
      secret = process.env.ABACATEPAY_SECRET ?? process.env.ABACATEPAY_API_KEY
    } = this.options;
    if (!secret)
      throw new AbacatePayError("We could not find any AbacatePay secret, use REST({ secret })");
    return {
      "Content-Type": "application/json",
      Authorization: `Bearer ${secret}`,
      ...this.options.headers,
      ...custom
    };
  }
}
// ../types/types/version.ts
var API_VERSION = "2";

// ../types/types/index.ts
var API_BASE_URL = "https://api.abacatepay.com/";
// src/version.ts
var version2 = "1.2.0";
// ../types/types/v2/routes.ts
var Routes = {
  customers: {
    create: "/customers/create",
    list({ page = 1, limit = 20 } = {}) {
      return `/customers/list?page=${page}&limit=${limit}`;
    },
    get(id) {
      return `/customers/get?id=${id}`;
    },
    delete: "/customers/delete"
  },
  checkouts: {
    create: "/checkouts/create",
    list: "/checkouts/list",
    get(id) {
      return `/checkouts/get?id=${id}`;
    }
  },
  transparents: {
    createQRCode: "/transparents/create",
    simulatePayment(id) {
      return `/transparents/simulate-payment?id=${id}`;
    },
    checkStatus(id) {
      return `/transparents/check?id=${id}`;
    }
  },
  coupons: {
    create: "/coupons/create",
    list({ page = 1, limit = 20 } = {}) {
      return `/coupons/list?page=${page}&limit=${limit}`;
    },
    get(id) {
      return `/coupons/get?id=${id}`;
    },
    delete: "/coupons/delete",
    toggleStatus: "/coupons/toggle"
  },
  payouts: {
    create: "/payouts/create",
    get(externalId) {
      return `/payouts/get?externalId=${externalId}`;
    },
    list({ page = 1, limit = 20 } = {}) {
      return `/payouts/list?page=${page}&limit=${limit}`;
    }
  },
  store: {
    get: "/store/get"
  },
  mrr: {
    get: "/public-mrr/mrr",
    merchant: "/public-mrr/merchant-info",
    revenue(start, end) {
      return `/public-mrr/revenue?startDate=${start}&endDate=${end}`;
    }
  },
  subscriptions: {
    create: "/subscriptions/create",
    list({ cursor, limit = 20 } = {}) {
      const query = new URLSearchParams({ limit: `${limit}` });
      if (cursor)
        query.append("cursor", cursor);
      return `/subscriptions/list?${query}`;
    }
  },
  products: {
    create: "/products/create",
    list({ page = 1, limit = 20 } = {}) {
      return `/products/list?page=${page}&limit=${limit}`;
    },
    get({ id, externalId } = {}) {
      const query = new URLSearchParams;
      if (id)
        query.append("id", id);
      if (externalId)
        query.append("externalId", externalId);
      return `/products/get?${query}`;
    }
  }
};
// src/v2/index.ts
var AbacatePay = ({ secret, rest: rest2 }) => {
  const client = new REST({
    secret,
    ...rest2,
    version: 2
  });
  return {
    rest: client,
    customers: {
      get(id) {
        return client.get(Routes.customers.get(id));
      },
      delete(id) {
        return client.delete(Routes.customers.delete, {
          body: { id }
        });
      },
      create(body) {
        return client.post(Routes.customers.create, { body });
      },
      list(query) {
        return client.get(Routes.customers.list(query));
      }
    },
    checkouts: {
      create(body) {
        return client.post(Routes.checkouts.create, { body });
      },
      list() {
        return client.post(Routes.checkouts.list);
      },
      get(id) {
        return client.get(Routes.checkouts.get(id));
      }
    },
    pix: {
      create(body) {
        return client.post(Routes.transparents.createQRCode, { body });
      },
      simulate(id, metadata) {
        return client.post(Routes.transparents.simulatePayment(id), { body: { metadata } });
      },
      status(id) {
        return client.get(Routes.transparents.checkStatus(id));
      }
    },
    coupons: {
      create(body) {
        return client.post(Routes.coupons.create, {
          body
        });
      },
      delete(id) {
        return client.delete(Routes.coupons.delete, {
          body: { id }
        });
      },
      get(id) {
        return client.get(Routes.coupons.get(id));
      },
      list(query) {
        return client.get(Routes.coupons.list(query));
      },
      toggleStatus(id) {
        return client.patch(Routes.coupons.toggleStatus, { body: { id } });
      }
    },
    store: {
      get() {
        return client.get(Routes.store.get);
      }
    },
    mrr: {
      get() {
        return client.get(Routes.mrr.get);
      },
      revenue({ startDate, endDate }) {
        return client.get(Routes.mrr.revenue(startDate, endDate));
      },
      merchant() {
        return client.get(Routes.mrr.merchant);
      }
    },
    payouts: {
      create(body) {
        return client.post(Routes.payouts.create, { body });
      },
      get(id) {
        return client.get(Routes.payouts.get(id));
      },
      list(query) {
        return client.get(Routes.payouts.list(query));
      }
    },
    subscriptions: {
      create(body) {
        return client.post(Routes.subscriptions.create, { body });
      },
      list(query) {
        return client.get(Routes.subscriptions.list(query));
      }
    },
    products: {
      create(body) {
        return client.post(Routes.products.create, {
          body
        });
      },
      get(query) {
        return client.get(Routes.products.get(query));
      },
      list(query) {
        return client.get(Routes.products.list(query));
      }
    }
  };
};

// src/index.ts
var ABACATEPAY_DOCS_URL = "https://docs.abacatepay.com/";
export {
  version2 as version,
  HTTPError,
  AbacatePayError,
  AbacatePay,
  API_VERSION,
  API_BASE_URL,
  ABACATEPAY_DOCS_URL
};
