// @abacatepay/types/dist/v2/routes.js
export const Routes = {
    customers: {
        /**
         * POST - https://api.abacatepay.com/v2/customers/create
         */
        create: '/customers/create',
        /**
         * GET - https://api.abacatepay.com/v2/customers/list
         */
        list({ page = 1, limit = 20 } = {}) {
            return `/customers/list?page=${page}&limit=${limit}`;
        },
        /**
         * GET - https://api.abacatepay.com/v2/customers/get
         */
        get(id) {
            return `/customers/get?id=${id}`;
        },
        /**
         * DELETE - https://api.abacatepay.com/v2/customers/delete
         */
        delete: '/customers/delete',
    },
    checkouts: {
        /**
         * POST - https://api.abacatepay.com/v2/checkouts/create
         */
        create: '/checkouts/create',
        /**
         * GET - https://api.abacatepay.com/v2/checkouts/list
         */
        list: '/checkouts/list',
        /**
         * GET - https://api.abacatepay.com/v2/checkouts/get
         */
        get(id) {
            return `/checkouts/get?id=${id}`;
        },
    },
    transparents: {
        /**
         * POST - https://api.abacatepay.com/v2/transparents/create
         */
        createQRCode: '/transparents/create',
        /**
         * POST - https://api.abacatepay.com/v2/transparents/simulate-payment
         */
        simulatePayment(id) {
            return `/transparents/simulate-payment?id=${id}`;
        },
        /**
         * GET - https://api.abacatepay.com/v2/transparents/check
         */
        checkStatus(id) {
            return `/transparents/check?id=${id}`;
        },
    },
    coupons: {
        /**
         * POST - https://api.abacatepay.com/v2/coupons/create
         */
        create: '/coupons/create',
        /**
         * GET - https://api.abacatepay.com/v2/coupons/list
         */
        list({ page = 1, limit = 20 } = {}) {
            return `/coupons/list?page=${page}&limit=${limit}`;
        },
        /**
         * GET - https://api.abacatepay.com/v2/coupons/get
         */
        get(id) {
            return `/coupons/get?id=${id}`;
        },
        /**
         * DELETE - https://api.abacatepay.com/v2/coupons/delete
         */
        delete: '/coupons/delete',
        /**
         * PATCH - https://api.abacatepay.com/v2/coupons/toggle
         */
        toggleStatus: '/coupons/toggle',
    },
    payouts: {
        /**
         * POST - https://api.abacatepay.com/v2/payouts/create
         */
        create: '/payouts/create',
        /**
         * GET - https://api.abacatepay.com/v2/payouts/get
         */
        get(externalId) {
            return `/payouts/get?externalId=${externalId}`;
        },
        /**
         * GET - https://api.abacatepay.com/v2/payouts/list
         */
        list({ page = 1, limit = 20 } = {}) {
            return `/payouts/list?page=${page}&limit=${limit}`;
        },
    },
    store: {
        /**
         * GET - https://api.abacatepay.com/v2/store/get
         */
        get: '/store/get',
    },
    mrr: {
        /**
         * GET - https://api.abacatepay.com/v2/public-mrr/mrr
         */
        get: '/public-mrr/mrr',
        /**
         * GET - https://api.abacatepay.com/v2/public-mrr/merchant-info
         */
        merchant: '/public-mrr/merchant-info',
        /**
         * GET - https://api.abacatepay.com/v2/public-mrr/renevue
         */
        revenue(start, end) {
            return `/public-mrr/revenue?startDate=${start}&endDate=${end}`;
        },
    },
    subscriptions: {
        /**
         * POST - https://api.abacatepay.com/v2/subscriptions/create
         */
        create: '/subscriptions/create',
        /**
         * GET - https://api.abacatepay.com/v2/subscriptions/list
         */
        list({ cursor, limit = 20 } = {}) {
            const query = new URLSearchParams({ limit: `${limit}` });
            if (cursor)
                query.append('cursor', cursor);
            return `/subscriptions/list?${query}`;
        },
    },
    products: {
        /**
         * POST - https://api.abacatepay.com/v2/products/create
         */
        create: '/products/create',
        /**
         * GET - https://api.abacatepay.com/v2/products/list
         */
        list({ page = 1, limit = 20 } = {}) {
            return `/products/list?page=${page}&limit=${limit}`;
        },
        /**
         * GET - https://api.abacatepay.com/v2/products/get
         */
        get({ id, externalId } = {}) {
            const query = new URLSearchParams();
            if (id)
                query.append('id', id);
            if (externalId)
                query.append('externalId', externalId);
            return `/products/get?${query}`;
        },
    },
};
